SpecialContentStaging.php

```diff
@@ -18,138 +19,319 @@ function __construct() {
+	 * @return WikiPage
+	 * @throws MWException
+	 */
+	private function copyPage( $prefix, WikiPage $sourcePage, $source, $target ) {
+		$titleSrc = $sourcePage->getTitle()->mTextform;
+		if ( $source === '' ) {
+			$titleTarget = $this->mwNamespace . str_replace( $prefix . '/', $prefix . '/' . $target . '/', $titleSrc );
 		} else {
-			$titleTarget = $this->mwNamespace . str_replace( $prefix . "/" . $source, $prefix . "/" . $target, $titleSrc );
+			$titleTarget = $this->mwNamespace . str_replace( $prefix . '/' . $source, $prefix . '/' . $target, $titleSrc );
 		}
-		
-		$pageContent = $objSrc->getText();
+
+		$pageContent = $sourcePage->getContent()->getNativeData();
 		$pageContent = $this->replaceStageInternalRefs( $prefix, $pageContent, $source, $target );
 
-		$objTarget = WikiPage::factory ( Title::newFromText( $titleTarget ) );
-		$objTarget->doEdit( $pageContent, "Staging content from " . $source . " to " . $target );
-		
-		return $objTarget;
+		$targetPage = WikiPage::factory( Title::newFromText( $titleTarget ) );
+		$targetPage->doEditContent( new WikitextContent( $pageContent ), 'Staging content from ' . $source . ' to ' . $target );
+
+		return $targetPage;
+	}
+
+	/**
+	 * @param WikiPage $sourceWikiPage
+	 * @param WikiPage $targetWikiPage
+	 * @param string $sourceStage
+	 * @param string $targetStage
+	 * @return bool
+	 */
+	private function stageContentDiffers( WikiPage $sourceWikiPage, WikiPage $targetWikiPage, $sourceStage, $targetStage ) {
+		$sourceText = $sourceWikiPage->getContent()->getNativeData();
+		$targetText = $targetWikiPage->getContent()->getNativeData();
+
+		return $this->replaceStageInternalRefs( $this->pagePrefix, $sourceText, $sourceStage, $targetStage ) !== $targetText;
+	}
+
+	/**
+	 * @param $wikiPage
+	 * @return bool
+	 */
+	private function wikiPageExists( $wikiPage ) {
+		return get_class( $wikiPage ) === 'WikiPage';
+	}
+
+	/**
+	 * @param WikiPage $wikiPage
+	 * @param bool $showArchived
+	 * @return bool
+	 */
+	private function shouldPageShow( WikiPage $wikiPage, $showArchived ) {
+		return $this->isArchivedPage( $wikiPage ) xor !$showArchived;
+	}
+
+	/**
+	 * @param string $title
+	 * @throws MWException
+	 */
+	private function archivePage( $title ) {
+		foreach ( $this->stages as $stage => $number ) {
+			$archiveTitle = $this->mwNamespace . $this->pagePrefix . '/' . $stage . '/' . $title;
+			$archivePage = WikiPage::factory( Title::newFromText( $archiveTitle ) );
+			$this->doArchivePage( $archivePage );
+		}
+	}
+
+	/**
+	 * @param WikiPage $page
+	 * @return bool
+	 * @throws MWException
+	 */
+	private function doArchivePage( WikiPage $page ) {
+		$oldContent = $page->getContent();
+
+		if ( $oldContent === null || $this->isArchivedPage( $page ) ) {
+			return false;
+		}
+
+		$text = $oldContent->getNativeData();
+
+		$text .= "\n[[Category:ContentStagingArchive]]";
+		$page->doEditContent( new WikitextContent( $text ), User::newFromSession(), 'archived by ContentStaging' );
```

Comment:
Yet another thing that I've missed previously: this call is not setting the summary correctly, should be something like `$page->doEditContent( new WikitextContent( $text ), 'archived by ContentStaging', 0, false, User::newFromSession() );